lib/SIL/SILInstructions.cpp

```diff
@@ -31,6 +31,86 @@
 using namespace swift;
 using namespace Lowering;
 
+// Collect used open archetypes from a given type.
+static void
+collectOpenedArchetypes(CanType Ty,
+                        SmallVectorImpl<CanType> &openedArchetypes) {
+  if (!Ty || !Ty->hasOpenedExistential())
+    return;
+  Ty.visit([&](Type t) {
+    if (t->isOpenedExistential()) {
+      // Add this opened archetype if it was not seen yet.
+      // We don't use a set here, because the number of open archetypes
+      // is usually very small and using a real set may introduce too
+      // much overhead.
+      if (std::find(openedArchetypes.begin(), openedArchetypes.end(),
+                    t.getCanonicalTypeOrNull()) == openedArchetypes.end())
```

Comment:
Oh, I see, you're worried about the overhead of using a set.  LLVM has a SmallSetVector, although now that I look at it, it looks like a pretty bad data structure.